### DISCUSSION TITLE
Lesson 9: (Introduction to passing private keys as vm.envUint) Encountering error "Failed Parsing $PRIVATE_KEY as UINT256 type"

### DISCUSSION THREAD

**Discussion by DanCsoftware** [Cyfrin/foundry-full-course-cu#1525] [Q&A] [answered]
Hey guys, been a bit confused with the following error I've been receiving when trying to do a fork test on Sepolia network, here is the error I'm receiving.

**(Error log)**

No files changed, compilation skipped
2024-03-25T07:57:15.692896Z ERROR forge::runner: setUp failed reason=failed parsing $PRIVATE_KEY as type `uint256`: parser error:
$PRIVATE_KEY
                                                                  ^
 contract=0x34A1D3fff3958843C43aD80F30b94c510645C316

Running 1 test for test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: setup failed: failed parsing $PRIVATE_KEY as type `uint256`: parser error:
$PRIVATE_KEY
                                                                  ^
] setUp() (gas: 0)
Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 3.43s
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: setup failed: failed parsing $PRIVATE_KEY as type `uint256`: parser error:
$PRIVATE_KEY
                                                                  ^
] setUp() (gas: 0)

Encountered a total of 1 failing tests, 0 tests succeeded

**(End of Error log)**

However, looking through the discussion forums people were stating it was because my Private key in my .env file wasn't formatted correctly and to add an "0x" to it, but either way i'm still encountering the same error, I'm making sure to .source env after my changes and ECHO to ensure it's updated correctly but I've been encountering the same error each time.

Here is my helper config code

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

contract HelperConfig is Script {
    struct NetworkConfig {
        uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint64 subscriptionId;
        uint32 callBackGasLimit;
        address link;
        uint256 deployerKey;
    }

    uint256 public constant DEFAULT_ANVIL_KEY =
        0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;
    NetworkConfig public activeNetworkConfig;

    constructor() {
        if (block.chainid == 11155111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else {
            activeNetworkConfig = getAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public view returns (NetworkConfig memory) {
        return
            NetworkConfig({
                entranceFee: 0.01 ether,
                interval: 30,
                vrfCoordinator: 0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625,
                gasLane: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
                subscriptionId: 5613, //update with our subId
                callBackGasLimit: 500000, // 500,000 gas
                link: 0x779877A7B0D9E8603169DdbD7836e478b4624789,
                deployerKey: vm.envUint("PRIVATE_KEY")
            });
    }

    function getAnvilEthConfig() public returns (NetworkConfig memory) {
        if (activeNetworkConfig.vrfCoordinator != address(0)) {
            return activeNetworkConfig;
        }
        uint96 baseFee = 0.25 ether; // 0.25 link
        uint96 gasPriceLink = 1e9; // 1 gwei link

        vm.startBroadcast();
        VRFCoordinatorV2Mock vrfCoordinatorMock = new VRFCoordinatorV2Mock(
            baseFee,
            gasPriceLink
        );
        LinkToken link = new LinkToken();
        vm.stopBroadcast();
        return
            NetworkConfig({
                entranceFee: 0.01 ether,
                interval: 30,
                vrfCoordinator: address(vrfCoordinatorMock),
                gasLane: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
                subscriptionId: 0, //our script will add this
                callBackGasLimit: 500000, // 500,000 gas
                link: address(link),
                deployerKey: DEFAULT_ANVIL_KEY
            });
    }
}
```

Appreciate the help!

**Comment by EngrPips** [comment]
Hello @DanCsoftware , prepend the private key in your `.env` file with `0x` and let me know if that works for you. basically if you initially have 

`$PRIVATE_KEY=2437893780879979003890`

after prepending it with `0x` you should now have the below 

`$PRIVATE_KEY=0x2437893780879979003890`

**Reply by DanCsoftware** [reply]
@EngrPips Hmm it still is giving me the same error, I made sure to source .env and ECHO my private key to ensure the change has been saved to the file, it's still giving me the same failed parsing type error.

**Reply by EngrPips** [reply]
Please push your code to github and share me the link to it, so that i can clone it and set it up locally on my machine to try debug the issue.

**Reply by DanCsoftware** [reply]
Absolutely, https://github.com/DanCsoftware/foundry-smart-contract-lottery-f24/tree/main/foundry-smart-contract-lottery-f23, thanks again for the help!

**Reply by EngrPips** [reply]
Anytime friend, thanks for sharing i will get back with whatever i find

**Reply by EngrPips** [reply]
Hello @DanCsoftware , i couldn't clone the repo because it is a sub-repo of some main repo, could you please push the project directory separately to a github repo of it's own 

<img width="1094" alt="Screenshot 2024-03-25 at 1 56 36 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/d01679bf-4c4f-4890-98cd-d47c99103e91">

**Reply by DanCsoftware** [reply]
Oh okay, sorry I'm still unfamiliar with github, did it work now? https://github.com/DanCsoftware/foundry-smart-contract-lottery-f23

**Reply by EngrPips** [reply]
Yeah this works, thanks

**Reply by DanCsoftware** [reply]
So I did some digging and actually found the error, though it was quite silly. In my ENV file there was a ";" at the end of each address, I think when it was reading the sepolia rpc url it was also taking that address and popping some error. 

However it led my fork-url to just one more error and I'm not sure why.

Failing tests:
```
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: Error != expected error: 0x584327aa00000000000000000000000000000000000000000000000004f3c18bfe17f90000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 != ] testPerformUpkeepRevertIfCheckUpkeepIsFalse() (gas: 17308)
```

Would you happen to know why? I double checked my testperformupkeeprevertifcheckupkeepisfalse function and it looks correct on my end. 

```
`    function testPerformUpkeepRevertIfCheckUpkeepIsFalse() public {
        uint256 currentBalance = 0;
        uint256 numPlayers = 0;
        uint256 raffleState = 0; // we know based on our numerical values on what open and closed mean
        vm.expectRevert(
            abi.encodeWithSelector(
                Raffle.Raffle__UpkeepNotNeeded.selector,
                currentBalance,
                numPlayers,
                raffleState
            )
        );
        raffle.performUpkeep("");`
```

**Reply by EngrPips** [reply, marked as answer]
Hello @DanCsoftware , The easiest way to fix this error would be for you set your `uint256 currentBalance` as i do below 

```javascript
uint256 currentBalance = address(raffle).balance
```
Like that your test should pass

**Reply by DanCsoftware** [reply]
Thanks a ton for the help, @EngrPips!

**Reply by EngrPips** [reply]
Anytime buddy, i am glad i am  at least of use to someone
